File: 17.Backend Security - Everything You Need to Know/code/typescript/complete_secure_login_endpoint_9.ts
import { Request, Response } from "express";
import crypto from "crypto";
import { Pool } from "pg";
import { Redis } from "ioredis";

// verifyArgon2 mock
const verifyArgon2 = async (password: string, hash: string) => true;

export function loginHandler(db: Pool, redis: Redis) {
    return async (req: Request, res: Response) => {
        const { email, password } = req.body;

        // 1. Validate format (first line of defence)
        if (!email || !email.includes("@") || !password || password.length < 8) {
            res.status(400).send("invalid credentials");
            return;
        }

        try {
            // 2. Parameterised query, no SQL injection possible
            const result = await db.query(
                "SELECT id, password_hash FROM users WHERE email = $1",
                [email]
            );

            if (result.rows.length === 0) {
                res.status(401).send("invalid email or password");
                return;
            }

            const { id: userID, password_hash: hashedPass } = result.rows[0];

            // 3. Generic error, never reveal whether email exists
            if (!(await verifyArgon2(password, hashedPass))) {
                res.status(401).send("invalid email or password");
                return;
            }

            // 4. Cryptographically secure session ID
            const sessionID = crypto.randomBytes(32).toString("base64url");

            // 5. Store session in Redis with metadata
            await redis.set(`session:${sessionID}`, userID, "EX", 7 * 24 * 3600);

            // 6. Secure cookie, HttpOnly, Secure, SameSite=Strict
            res.cookie("session_id", sessionID, {
                httpOnly: true,
                secure: true,
                sameSite: "strict",
                maxAge: 7 * 24 * 3600 * 1000, // ms
            });
            res.status(200).send("OK");
        } catch (err) {
            res.status(500).send("server error");
        }
    };
}

File: 17.Backend Security - Everything You Need to Know/code/rust/complete_secure_login_endpoint_9.rs
use actix_web::{web, HttpResponse, Responder};
use actix_web::cookie::{Cookie, SameSite};
use sqlx::PgPool;
use redis::AsyncCommands;
use serde::Deserialize;
use rand::RngCore;
use base64::{Engine as _, engine::general_purpose::URL_SAFE_NO_PAD};

#[derive(Deserialize)]
pub struct LoginRequest {
    email: String,
    password: String,
}

// Mock
fn verify_argon2(_password: &str, _hash: &str) -> bool { true }

pub async fn login_handler(
    req: web::Json<LoginRequest>,
    db: web::Data<PgPool>,
    redis: web::Data<redis::Client>,
) -> impl Responder {
    // 1. Validate format (first line of defence)
    if !req.email.contains('@') || req.password.len() < 8 {
        return HttpResponse::BadRequest().body("invalid credentials");
    }

    // 2. Parameterised query, no SQL injection possible
    let record = match sqlx::query!(
        "SELECT id, password_hash FROM users WHERE email = $1",
        req.email
    )
    .fetch_optional(db.get_ref())
    .await {
        Ok(Some(r)) => r,
        _ => {
            // 3. Generic error, never reveal whether email exists
            return HttpResponse::Unauthorized().body("invalid email or password");
        }
    };

    // 3. Generic error, never reveal whether email exists
    if !verify_argon2(&req.password, &record.password_hash) {
        return HttpResponse::Unauthorized().body("invalid email or password");
    }

    // 4. Cryptographically secure session ID
    let mut b = vec![0u8; 32];
    rand::rngs::OsRng.fill_bytes(&mut b);
    let session_id = URL_SAFE_NO_PAD.encode(&b);

    // 5. Store session in Redis with metadata
    let mut con = match redis.get_multiplexed_async_connection().await {
        Ok(c) => c,
        Err(_) => return HttpResponse::InternalServerError().body("server error"),
    };
    let _: () = match con.set_ex(format!("session:{}", session_id), record.id.to_string(), 7 * 24 * 3600).await {
        Ok(_) => (),
        Err(_) => return HttpResponse::InternalServerError().body("server error"),
    };

    // 6. Secure cookie, HttpOnly, Secure, SameSite=Strict
    let cookie = Cookie::build("session_id", session_id)
        .http_only(true)
        .secure(true)
        .same_site(SameSite::Strict)
        .max_age(time::Duration::days(7))
        .finish();

    HttpResponse::Ok().cookie(cookie).finish()
}
